crates/consensus/src/transaction/envelope.rs

```diff
@@ -245,6 +247,92 @@ impl Encodable2718 for TxEnvelope {
     }
 }
 
+impl TryFrom<Transaction> for TxEnvelope {
+    type Error = Eip2718Error;
+
+    fn try_from(tx: Transaction) -> Result<Self, Self::Error> {
+        let signature = tx
+            .signature
+            .ok_or(Eip2718Error::MissingSignature)?
+            .try_into()
+            .map_err(Eip2718Error::SignatureError)?;
+
+        match tx.transaction_type.map(|t| t.to()) {
```

Comment:
each match arm should delegate to another TryFrom impl

Reply:
should not match on integers, instead de-duplicate the `TransactionType` enums from consensus and rpc-types

Reply:
tx.transaction_type is `Option<U64>`:
https://github.com/alloy-rs/alloy/blob/218ece4ca60698c04f6fe29eda629add642aeebd/crates/rpc-types/src/eth/transaction/mod.rs#L76-L81

would it make sense to add `impl TryFrom<Option<U64>> for TxType`?

Reply:
missing type always implies 0, right? so you can use `unwrap_or_default()` on the option before converting.

Reply:
yep, but 0 seems to be treated as invalid type for this conversion

https://github.com/alloy-rs/alloy/blob/218ece4ca60698c04f6fe29eda629add642aeebd/crates/consensus/src/transaction/envelope.rs#L40-L51

Reply:
cc @Rjected some tools use `type: 0` to reflect legacy txns. shouldn't we support this?

Reply:
> cc @Rjected some tools use `type: 0` to reflect legacy txns. shouldn't we support this?

yes, just need to make sure that we still disallow decode attempts for "type 0" 2718 txs. But this can be done in decoding and not here, so try_from can support type 0

Reply:
dope. so this `try_from` can be updated because there is a corresponding check in the `impl Decodable2718 for TxEnvelope` match arm for the legacy type 👍